sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/AliasAwareOutputExpression.scala

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.plans
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.catalyst.expressions.{Alias, Attribute, AttributeSet, Empty2Null, Expression, NamedExpression, SortOrder}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A trait that provides functionality to handle aliases in the `outputExpressions`.
+ */
+trait AliasAwareOutputExpression extends SQLConfHelper {
+  protected val aliasCandidateLimit = conf.getConf(SQLConf.EXPRESSION_PROJECTION_CANDIDATE_LIMIT)
+  protected def outputExpressions: Seq[NamedExpression]
+  /**
+   * This method can be used to strip expression which does not affect the result, for example:
+   * strip the expression which is ordering agnostic for output ordering.
+   */
+  protected def strip(expr: Expression): Expression = expr
+
+  // Build an `Expression` -> `Attribute` alias map.
+  // There can be multiple alias defined for the same expressions but it doesn't make sense to store
+  // more than `aliasCandidateLimit` attributes for an expression. In those cases the old logic
+  // handled only the last alias so we need to make sure that we give precedence to that.
+  // If the `outputExpressions` contain simple attributes we need to add those too to the map.
+  private lazy val aliasMap = {
+    val aliases = mutable.Map[Expression, mutable.ArrayBuffer[Attribute]]()
+    outputExpressions.reverse.foreach {
+      case a @ Alias(child, _) =>
+        val buffer = aliases.getOrElseUpdate(strip(child.canonicalized), mutable.ArrayBuffer.empty)
+        if (buffer.size < aliasCandidateLimit) {
+          buffer += a.toAttribute
+        }
+      case _ =>
+    }
+    outputExpressions.foreach {
+      case a: Attribute if aliases.contains(a.canonicalized) =>
```

Comment:
What does this do ? do you mean  `!aliases.contains(a.canonicalized)` ?

Reply:
If we have `c, c AS a` projection then we need to add both the original `c` attribute and `a` to the alternatives of `a`. But we don't need to add an attribute if it isn't aliased.

Reply:
what's the behavior of `c as a` ? this code seems to return both `c` and `a`. I think the right way should be
`if AttributSet(outputExpressions).contains(a) => // add a to buffer`

Reply:
This case doesn't match `c AS a`. This case makes sure that if `c` -> `a` has been added to the map by the previous forach then `c` -> `c` should be added too.

I think your code would add all attributes, but that is not needed.

Reply:
why `c -> c` should be added ? the outputExpression only contains `c as a`, shall we only return `c -> a` ?

Reply:
I see your point, but this code show me that, it will return both `c -> c` and `c -> a` even the outputExpression is `c as a`. no ?

```scala
// outputExpressions is `c as a`
outputExpressions.foreach {
  // aliases is `c -> a` in first foreach so it's true
  case a: Attribute if aliases.contains(a.canonicalized) =>
    val buffer = aliases(a.canonicalized)
    if (buffer.size < aliasCandidateLimit) {
      // c is added into buffer
      buffer += a
    }
```

Reply:
No, why would `c AS a` match on `case a: Attribute`? The foreach runs on a `Seq[Expression]` not on an `Expression`.

Reply:
ah, I see it. how about this case: `c1, c2 as x` ? We also should add c1 into aliasMap right ?

Reply:
No. If `c1` is not aliased otherwise then we don't need to add it to the map. If the map doesn't contain any alias for an expression then the transformation does nothing with that `c1` attribute, just leave it in the expression tree as it is...

Reply:
But the aliasMap is not empty due to `c2 as x`. for this case, how can we preserve c1 if c1 does not add into aliasMap ?

Reply:
But the aliasMap is not empty due to c2 as x. for this case, how can we preserve c1 if c1 does not add into aliasMap ?

Reply:
oh I see it, nvm. @peter-toth  Thank you for the patience !

Reply:
Np, thanks for reviewing this PR @ulysses-you!